### DISCUSSION TITLE
[Idea] Refactor Connection Class

### DISCUSSION THREAD

**Discussion by ChillCode** [PrestaShop/PrestaShop#30879] [Ideas]
### The problem

Some customers got 1.000/10.000 fake connections per day when running **PrestaShop** on  production because of the Connection class.

So stats on Adminstats are completely useless, but after we installed a firewall module the connections drops just to 70-100 per day and have the same or even more placed orders.

### The solution

block or check crawlers/bots/spiders/scanners can be done using .htaccess but if there is a native class already would be better to use it.

### Alternatives

Add a native Firewall module to handle all bot connections, by now we are implementing a [mmdb ](https://metacpan.org/pod/MaxMind::DB::Writer) file to handle this without using DB to speed up the process.

The next function is part of the Firewall that check if Pinterest bot is legitim, and the result is cached on DB:

```
    /**
     * Check if UA and IP are Pinterest Bot valid.
     * @param string $userAgent User Agent.
     * @param int $ipAddress IP Address.
     * @return bool
     * @throws BotException
     */
    public static function isPinterestBot(string $userAgent, int $ipAddress)
    {
        if (!empty($userAgent) && preg_match('/ . static::$pinterestBot . /i', $userAgent)) {
            if ($ipAddress >= 921436417 && $ipAddress <= 921436671) {
                return true;
            } else {
                throw new BotException('[PrestaShop Firewall] Pinterest fake bot agent detected');
            }
        }

        return false;
    }
```

### Additional context

Those bots are not present on Connection Class and are shown as legitimate customers connections:

`Applebot`

https://support.apple.com/en-us/HT204683

`Bingbot|AdIdxBot`

https://www.bing.com/webmasters/help/which-crawlers-does-bing-use-8c184ec0
https://www.bing.com/toolbox/bingbot.json

`APIs-Google|AdsBot-Google-Mobile|AdsBot-Google|Mediapartners-Google|AdsBot-Google-Mobile-Apps|FeedFetcher-Google|Google-Read-Aloud|DuplexWeb-Google|googleweblight|Storebot-Google|Google-Site-Verification`

https://developers.google.com/search/docs/crawling-indexing/overview-google-crawlers
https://developers.google.com/static/search/apis/ipranges/googlebot.json

`Pinterest`

https://help.pinterest.com/es/business/article/pinterest-crawler

`DuckDuckBot`

https://help.duckduckgo.com/duckduckgo-help-pages/results/duckduckbot/

`facebookexternalhit|facebookcatalog`

https://developers.facebook.com/docs/sharing/webmasters/crawler

Also some paid services that can make 10.000 connections per hour if someone put the url on these sites:

`AhrefsBot|BLEXBot|Barkrowler|Cloudfind|Domains Project|DotBot|mj12bot|InternetMeasurement|IonCrawl|Rogerbot|SeekportBot|SemrushBot|serpstatbot|ZoominfoBot|Yellowbrandprotectionbot`

Or some web scanners:

`censys|foregenix|masscan|paloaltonetworks.com|sindresorhus`

I know there are thousands but since Symfony have a Crawler detector can be another alternative.

### Do you plan to work on this subject?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by kpodemski** [comment]
Hi @ChillCode !

It's definitely a good idea! Would you be interested in contributing changes to the project? I can help you with the process :)

**Reply by ChillCode** [reply]
Hi @kpodemski 

Yes, I'm willing to contribute, already installed **PrestaShop**  dev 8.1 and will start from here.

I'd like to have some guidance to find the best approach, by now I created a module that is hooked on DispatcherBefore and works pretty well to block badbots.

**Reply by kpodemski** [reply]
You can contact me on [PrestaShop project Slack](https://www.prestashop-project.org/slack/), we can take it from there :)
